instructions/src/lib.rs

```diff
@@ -1,162 +1,1532 @@
+            MemSize::HalfWord => 0x08,
+            MemSize::Word => 0x00
+        }
+    }
+}
+
+pub enum Addressing {
+    Undef,
+    Abs,
+    Ind
+}
+
+impl AsRef<Addressing> for Addressing {
+    fn as_ref(&self) -> &Self {
+        self
+    }
+}
+
+impl<'a> From<&'a Addressing> for u8 {
+    fn from(addressing: &'a Addressing) -> u8 {
+        match *addressing {
+            Addressing::Undef => 0x00,
+            Addressing::Abs => 0x20,
+            Addressing::Ind => 0x40
+        }
+    }
+}
+
+pub enum AddressSource {
+    Register,
+    Immediate
+}
+
+impl AsRef<AddressSource> for AddressSource {
+    fn as_ref(&self) -> &Self {
+        self
+    }
+}
+
+impl<'a> From<&'a AddressSource> for u8 {
+    fn from(addressing: &'a AddressSource) -> u8 {
+        match *addressing {
+            AddressSource::Register => 0x61,
+            AddressSource::Immediate => 0x00
+        }
+    }
 }
 
-impl <'p> Operation<'p> {
+pub struct Store {
+    mem_size: MemSize,
+    address_source: AddressSource,
+    dst: u8,
+    src: u8,
+    offset: u16,
+    imm: u32
+}
 
-    fn new(program: &'p mut ProgramCodeBuilder, op: u8) -> Operation<'p> {
-        Operation {
-            program: program,
-            op: op,
+impl Store {
+    pub fn new(mem_size: MemSize, address_source: AddressSource) -> Self {
+        Store {
+            mem_size: mem_size,
+            address_source: address_source,
             dst: 0x00,
             src: 0x00,
             offset: 0x00_00,
-            immediate: 0x00_00_00_00
+            imm: 0x00_00_00_00
         }
     }
+}
+
+impl Instruction for Store {
+    fn opt_code_byte(&self) -> u8 {
+        let size: u8 = self.mem_size.as_ref().into();
+        let address_source: u8 = self.address_source.as_ref().into();
+        0x62 | size | address_source
+    }
+
+    fn dst(&self) -> u8 {
+        self.dst
+    }
+
+    fn src(&self) -> u8 {
+        self.src
+    }
+
+    fn offset(&self) -> u16 {
+        self.offset
+    }
+
+    fn imm(&self) -> u32 {
+        self.imm
+    }
 
-    pub fn dst(&'p mut self, dst_register: u8) -> &'p mut Operation<'p> {
-        self.dst = dst_register;
+    fn dst_reg(mut self, dst: u8) -> Self {
+        self.dst = dst;
         self
     }
 
-    pub fn src(&'p mut self, src_register: u8) -> &'p mut Operation<'p> {
-        self.src = src_register;
+    fn src_reg(mut self, src: u8) -> Self {
+        self.src = src;
         self
     }
 
-    pub fn offset(&'p mut self, offset: u16) -> &'p mut Operation<'p> {
+    fn offset_bytes(mut self, offset: u16) -> Self {
         self.offset = offset;
         self
     }
 
-    pub fn value(&'p mut self, immediate: u32) -> &'p mut Operation<'p> {
-        self.immediate = immediate;
+    fn immediate(mut self, imm: u32) -> Self {
+        self.imm = imm;
         self
     }
+}
+
+pub struct Jump {
+    cmp: Comparison,
+    src_bit: Source,
+    dst: u8,
+    src: u8,
+    offset: u16,
+    imm: u32
+}
```

Comment:
`imm` and `offset` are considered as signed. We'll have to fix it at some point if we use `ebpf::Insn`.